### PR TITLE
[KHWEBA-48] Transaction dialog

### PR DESCRIPTION
**Pull request by gift-a**
image

### CODE REVIEW

**Review by Vladyslav-Plakhuta — DISMISSED** [review]

**Review by OlgaOrlova — DISMISSED** [review]

File: index.html [hunk 1]
```diff
@@ -6,6 +6,8 @@
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
   <meta http-equiv="X-UA-Compatible" content="ie=edge">
   <title>BUXY 2</title>
+  <link rel='stylesheet' href='http://fonts.googleapis.com/css?family=Roboto:400,700'>
+  <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
```
Comment (ihor7016) [inline index.html]: please make it consistent -- first should come `rel` and then `href`

File: src/components/app/app.html [hunk 2]
```diff
@@ -1 +1,8 @@
-<h1>Hi!</h1>
\ No newline at end of file
+<div class="app">
+  <transaction-dialog show="$ctrl.transactionDialogShow"></transaction-dialog>
+  <md-button class="app__add-transaction-dialog-activation md-fab md-primary" aria-label="FAB" ng-click="$ctrl.onAddTransactionClick($event)">
```
Comment (ihor7016) [inline src/components/app/app.html]: please do not use such long class names; BTW I thought that this is just `transaction-dialog` without `add`

File: src/components/transaction-dialog/transaction-dialog.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    this.$mdDialog.show({
+      template,
+      controllerAs: "$ctrl",
+      controller: this.dialogController(),
+      parent: angular.element(document.body),
+      targetEvent: ev,
+      clickOutsideToClose: true
+    });
+  }
+  dialogController() {
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.js]: please extract this function from class; or think of another way

**Comment by ihor7016** [comment]
BTW -- please make dialog layout like in vanilla app

File: src/components/app/app.js [hunk 4]
```diff
@@ -1,6 +1,23 @@
...
-export const appModule = angular.module("appModule", []).component("app", {
-  template
-}).name;
+import ngMaterial from "angular-material";
+
+import { transactionDialogModule } from "../transaction-dialog/transaction-dialog";
+
+export class AppController {
+  constructor() {
+    this.transactionDialogOpen;
```
Comment (ihor7016) [inline src/components/app/app.js]: why do you need this?

File: src/components/app/app.js [hunk 5]
```diff
@@ -1,6 +1,23 @@
...
+  }
+
+  onAddTransactionClick(event) {
+    this.transactionDialogOpen = { event: event };
```
Comment (ihor7016) [inline src/components/app/app.js]: what this line does?

File: src/components/transaction-dialog/transaction-dialog.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            <md-radio-button value="+" class="md-primary">Income</md-radio-button>
+          </md-radio-group>
+        </md-input-container>
+        <span flex="10"></span>
+        <md-input-container class="date-picker" flex="45">
+          <label>Date of transaction</label>
+          <md-datepicker ng-model="$ctrl.date" md-placeholder="Enter date" md-open-on-focus></md-datepicker>
+        </md-input-container>
+      </div>
+      <div layout='row' layout-align="space-between center">
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.html]: please use double quotes

File: src/components/transaction-dialog/transaction-dialog.html [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        <md-input-container flex="45">
+          <label>Description</label>
+          <input ng-model="$ctrl.description" type="text">
+        </md-input-container>
+        <md-input-container flex="45">
+          <label>Amount</label>
+          <input ng-model="$ctrl.amount" type="text">
+        </md-input-container>
+      </div>
+      <div layout='row' layout-align="space-between center">
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.html]: double quotes

File: src/components/transaction-dialog/transaction-dialog.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    if (changes.isOpen.currentValue) {
+      this.showDialog(changes.isOpen.currentValue.event);
+    }
+  }
+
+  showDialog(ev) {
+    this.$mdDialog.show({
+      template,
+      controllerAs: "$ctrl",
+      controller: () => new DialogController(this),
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.js]: * you do not need to instantiate class here -- you just need to pass class 
* you need to use `locals` prop in show() to pass variables that you need to controller

Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: if I use locals - it is the same, as I use this as props. here is an example
image

Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: What do you mean "to pass class"?

File: src/components/app/app.js (hunk 5 above)
Comment (gift-a) [reply src/components/app/app.js]: every press on button sends message to dialog to open and sends events, from where should dialog be shown

File: src/components/app/app.js (hunk 4 above)
Comment (gift-a) [reply src/components/app/app.js]: if I removed this, dialog will be opened right after page loading

Comment (gift-a) [reply src/components/app/app.js]: sorry, yes. I it seems I don't need this. no problem anymore without `this.transactionDialogOpen;` in constructor

File: src/components/app/app.js [hunk 9]
```diff
@@ -1,6 +1,19 @@
...
 
-export const appModule = angular.module("appModule", []).component("app", {
-  template
-}).name;
+import ngMaterial from "angular-material";
+
+import { transactionDialogModule } from "../transaction-dialog/transaction-dialog";
+
+export class AppController {
+  onAddTransactionClick(event) {
```
Comment (ihor7016) [inline src/components/app/app.js]: it is a handler so -- 'handleAddTrans...'

File: src/components/app/app.html [hunk 10]
```diff
@@ -1 +1,8 @@
-<h1>Hi!</h1>
\ No newline at end of file
+<div class="app">
+  <transaction-dialog is-open="$ctrl.transactionDialogOpen"></transaction-dialog>
+  <md-button class="app__transaction-dialog-activation md-fab md-primary" aria-label="FAB" ng-click="$ctrl.onAddTransactionClick($event)">
```
Comment (ihor7016) [inline src/components/app/app.html]: please format this like 
```
<md-button 
  class="app__transaction-dialog-activation md-fab md-primary" 
  aria-label="FAB" 
  ng-click="$ctrl.onAddTransactionClick($event)">

File: src/components/transaction-dialog/transaction-dialog.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        $mdDialog: this.$mdDialog,
+        $filter: this.$filter,
+        tags: this.tags,
+        accounts: this.accounts
+      }
+    });
+  }
+}
+
+class DialogController {
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.js]: export class

File: src/components/transaction-dialog/transaction-dialog.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  showDialog(ev) {
+    this.$mdDialog.show({
+      template,
+      controllerAs: "$ctrl",
+      controller: DialogController,
+      parent: angular.element(document.body),
+      targetEvent: ev,
+      clickOutsideToClose: true,
+      locals: {
+        $mdDialog: this.$mdDialog,
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.js]: isn't `$mdDialog` injected without this?

File: src/components/app/app.js [hunk 13]
```diff
@@ -1,6 +1,19 @@
...
+    this.transactionDialogOpen = { event: event };
```
Comment (ihor7016) [inline src/components/app/app.js]: why this is object? what is event?

File: src/components/transaction-dialog/transaction-dialog.js (hunk 12 above)
Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: no

File: src/components/transaction-dialog/transaction-dialog.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  showDialog(ev) {
+    this.$mdDialog.show({
+      template,
+      controllerAs: "$ctrl",
+      controller: DialogController,
+      parent: angular.element(document.body),
+      targetEvent: ev,
```
Comment (ihor7016) [inline src/components/transaction-dialog/transaction-dialog.js]: do you need this target event?

File: src/components/app/app.js (hunk 13 above)
Comment (gift-a) [reply src/components/app/app.js]: event is a button click

File: src/components/transaction-dialog/transaction-dialog.js (hunk 14 above)
Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: it is a target from where dialog shows. For now it is for animation, but in future (in my opinion) I will need it to know about data to edit

Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: ok. maybe I will not need it in future. but md-material uses it for dialog animation

Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: image

File: src/components/app/app.js [hunk 15]
```diff
@@ -1,6 +1,21 @@
...
-}).name;
+import ngMaterial from "angular-material";
+import ngAnimate from "angular-animate";
+import ngAria from "angular-aria";
+
+import { transactionDialogModule } from "../transaction-dialog/transaction-dialog";
+
+export class AppController {
+  handleAddTransactionClick(event) {
+    this.transactionDialogOpen = { event: event };
```
Comment (ihor7016) [inline src/components/app/app.js]: please pass event as separate prop -- it is really confusing to see prop `is-open` on dialog and you pass event wrapper into an object there

File: src/components/transaction-dialog/transaction-dialog.js (hunk 14 above)
Comment (ihor7016) [reply src/components/transaction-dialog/transaction-dialog.js]: ok

File: src/components/transaction-dialog/transaction-dialog.js (hunk 12 above)
Comment (ihor7016) [reply src/components/transaction-dialog/transaction-dialog.js]: ok

Comment (gift-a) [reply src/components/transaction-dialog/transaction-dialog.js]: yes. tried another way, than last time

File: src/components/app/app.js [hunk 16]
```diff
@@ -1,6 +1,24 @@
...
+import ngAria from "angular-aria";
+
+import { transactionDialogModule } from "../transaction-dialog/transaction-dialog";
+
+export class AppController {
+  constructor() {
+    this.transactionDialogOpen = { isOpen: false };
+  }
+  handleAddTransactionClick(event) {
+    this.transactionDialogOpen = { isOpen: true, openEvent: event };
```
Comment (ihor7016) [inline src/components/app/app.js]: why do you need to group these things into one object? you can pass it as separate 2 props to component

File: src/components/app/app.js [hunk 17]
```diff
@@ -1,6 +1,26 @@
...
-}).name;
+import ngMaterial from "angular-material";
+import ngAnimate from "angular-animate";
+import ngAria from "angular-aria";
+
+import { transactionDialogModule } from "../transaction-dialog/transaction-dialog";
+
+export class AppController {
+  handleAddTransactionClick(event) {
+    this.transactionDialogOpen = true;
```
Comment (ihor7016) [inline src/components/app/app.js]: please start boolean variables from `is` `has` `should` -- `isTransactionDialogOpen`

**Review by ihor7016 — APPROVED** [review]

**Pull request merged**